3 setup database

haltaf19 opened the pull request:
# CHANGELOG: 
- Create database for CRUD operations
- Add unit tests


Authored by @StuffByLiang , @Fandrew44  and @haltaf19

StuffByLiang committed:
setup basic database

Co-authored-by: Liang Liu <stuffbyliang@gmail.com>
Co-authored-by: Hasan Altaf <hasanaltaf2001@gmail.com>
Co-authored-by: Andrew Han <ahgreeting@gmail.com>

haltaf19 committed:
fixed webbased bugs

haltaf19 committed:
Add crud methods for database 🏢🐶

Co-authored-by: Liang Liu <stuffbyliang@gmail.com>
Co-authored-by: Hasan Altaf <hasanaltaf2001@gmail.com>
Co-authored-by: Andrew Han <ahgreeting@gmail.com>

haltaf19 committed:
Add unit tests for CRUD operations in database. 💯🔥👌🏿

Co-authored-by: Liang Liu <stuffbyliang@gmail.com>
Co-authored-by: Hasan Altaf <hasanaltaf2001@gmail.com>
Co-authored-by: Andrew Han <ahgreeting@gmail.com>

Commit:
Merge master into 3-setup-database

Commit:
Separate tests that can be run using flutter test (automated) and tests that must be run using flutter run (manual) so that the GitHub actions script will not try run manual tests using flutter test command

Commit:
Merge branch 'master' into 3-setup-database

StuffByLiang commented on lib/models/task.dart:
```diff
@@ -0,0 +1,18 @@
+  final int id;
+  final String description;
+  final String category;
+  final int clock;
```
We should probably change clock to something more representative of time @haltaf19 @Fandrew44  LOL before merging

sarahdagger replied:
I realize we probably need to keep track of when the task was started and the length of the task (my bad - I think I just wrote one in the issue). Let's change `clock` to `startTime` and `stopTime` or something along those lines (as long as we can figure out when the task started and how long it took). You can probably still keep them as `int`s representing [milliseconds since epoch](https://api.dart.dev/stable/2.10.3/dart-core/DateTime/millisecondsSinceEpoch.html) (there also might be a better way to store them - this is just a suggestion).

Commit:
Fix typo in dart.yml

Commit:
Merge branch '3-setup-database' of https://github.com/ubclaunchpad/time-tracker into 3-setup-database

sarahdagger commented on lib/main.dart:
```diff
@@ -1,5 +1,13 @@
+void main() async {
+  var db = new DatabaseHelper();
+
+  // final task = Task(id: 0, description: 'Fido', category: 'Work', clock: 2001);
```
Let's remove these commented out debugging lines

sarahdagger commented on lib/database/database_helper.dart:
```diff
@@ -0,0 +1,152 @@
+
+  initDb() async {
+    WidgetsFlutterBinding.ensureInitialized();
+    // io.Directory documentsDirectory = await getApplicationDocumentsDirectory();
```
I think these commented-out lines are an old implementation? If so, let's remove them since the current implementation works now.

sarahdagger commented on lib/database/database_helper.dart:
```diff
@@ -0,0 +1,152 @@
+  }
+
+  // Define a function that inserts tasks into the database
+  Future<void> insertTask(Task task) async {
```
It might be difficult for the user to assign the passed in `task` with a unique `id` manually. Maybe instead of making the user of the `insertTask` function pass in a task with an `id`, could we use `AUTOINCREMENT` for this (ie. change `CREATE TABLE task(id INTEGER PRIMARY KEY, description TEXT, clock INTEGER, category TEXT)` to `CREATE TABLE task(id INTEGER PRIMARY KEY AUTOINCREMENT, description TEXT, clock INTEGER, category TEXT)`? Then, we could ignore the `id` of the passed in `task` and instead return the `id` assigned through `AUTOINCRMENT` so the user could have a reference to the `id` (this could either be done by simply returning the integer `id` or by returning the passed in `task` with the `id` field updated). If you have another idea to solve the problem of the manual assignment of a unique `id`, feel free to suggest it as well!

sarahdagger commented on lib/database/database_helper.dart:
```diff
@@ -0,0 +1,152 @@
+  }
+
+  // delete
+  Future<void> deleteTask(int id) async {
```
It might be useful to return whether or not the the deletion was successful. I believe that `await dbClient.delete(...` should return a count of the number of entries deleted from the database. Returning this count should be sufficient.

sarahdagger commented on lib/database/database_helper.dart:
```diff
@@ -0,0 +1,152 @@
+  }
+
+  // Update
+  Future<void> updateTask(Task task) async {
```
It might be useful to return whether or not the update was successful. I believe `await dbClient.update...` returns a count of the number of entries updated and returning this value should be sufficient.

sarahdagger commented on lib/database/database_helper.dart:
```diff
@@ -0,0 +1,152 @@
+    // `conflictAlgorithm` to use in case the same Task is inserted twice.
+    //
+    // In this case, replace any previous data.
+    await dbClient.insert(
```
I would suggest changing the conflict algorithm from `replace` to `fail` so that debugging is easier in the future.

sarahdagger reviewed (commented):
I left a couple of suggestions that I'd love if you'd take a look at (please feel free to discuss), but overall looks really good! I also merged master into this branch and made a couple of changes to the structure of the tests directory to make GitHub Actions happy, so make sure to pull before making additional changes. Thanks, guys!

Commit:
Put Task model and database package in src package

haltaf19 committed:
Clean database_helper and database_test 🗑
ID autoincrements rather than manual input🎈
Change the database from field "clock" to fields "startTime" and "stopTime"🕰

Co-authored-by: Liang Liu <stuffbyliang@gmail.com>
Co-authored-by: Hasan Altaf <hasanaltaf2001@gmail.com>
Co-authored-by: Andrew Han <ahgreeting@gmail.com>

Commit:
Merge branch 'master' into 3-setup-database
